examples/pytorch/gin/multi_gpu_graph_prediction.py

```diff
@@ -0,0 +1,151 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.distributed as dist
+import torch.optim as optim
+import dgl
+import dgl.nn as dglnn
+from dgl.data import AsGraphPredDataset
+from dgl.dataloading import GraphDataLoader as DataLoader
+from ogb.graphproppred import DglGraphPropPredDataset, Evaluator
+from ogb.graphproppred.mol_encoder import AtomEncoder, BondEncoder
+from tqdm import tqdm
+import time
+import argparse
+
+class MLP(nn.Module):
+    def __init__(self, in_feats):
+        super().__init__()
+        self.mlp = nn.Sequential(
+            nn.Linear(in_feats, 2 * in_feats),
+            nn.BatchNorm1d(2 * in_feats),
+            nn.ReLU(),
+            nn.Linear(2 * in_feats, in_feats),
+            nn.BatchNorm1d(in_feats)
+        )
+
+    def forward(self, h):
+        return self.mlp(h)
+
+class GIN(torch.nn.Module):
+    def __init__(self, n_hidden, n_output, n_layers=5):
+        super().__init__()
+        self.node_encoder = AtomEncoder(n_hidden)
+        self.edge_encoders = nn.ModuleList([
+            BondEncoder(n_hidden) for _ in range(n_layers)])
+
+        self.pool = dglnn.AvgPooling()
+        self.dropout = nn.Dropout(0.5)
+        self.layers = nn.ModuleList()
+        for _ in range(n_layers):
+            self.layers.append(dglnn.GINEConv(MLP(n_hidden), learn_eps=True))       
+        self.predictor = nn.Linear(n_hidden, n_output)    
+
+        # add virtual node
+        self.virtual_emb = nn.Embedding(1, n_hidden)
+        nn.init.constant_(self.virtual_emb.weight.data, 0)
+        self.virtual_layers = nn.ModuleList()
+        for _ in range(n_layers - 1):
+            self.virtual_layers.append(MLP(n_hidden))
+        self.virtual_pool = dglnn.SumPooling()
+
+    def forward(self, g, x, adj_t):
+        v_emb = self.virtual_emb.weight.expand(g.batch_size, -1)
+        hn = self.node_encoder(x)
+        for layer in range(len(self.layers)):
+            v_hn = dgl.broadcast_nodes(g, v_emb)
+            hn = hn + v_hn
+            he = self.edge_encoders[layer](adj_t)
+            hn = self.layers[layer](g, hn, he)
+            if layer != len(self.layers) - 1:
```

Comment:
L60 - L66 read a bit strange. Why dropout is needed for all the layers while all other non-linear transformation (e.g., activation and virtual nodes) are skipped for layers other than the last one? Can we simply have:
```python
if layer != len(self.layers) - 1:
    hn = F.relu(hn)
    hn = self.dropout(hn)
    v_emb_tmp = ...
    ...
```

Reply:
Yes, I think we can simplify the model to have Relu for every layer and keep v_emb excluded in the latest layer as it is used for next layer computation but it is already the last one . For relu, originally we followed the implementation [here](https://github.com/snap-stanford/ogb/blob/master/examples/graphproppred/mol/conv.py#L205), but according my test, it performs much the same, so it's ok to simplify the code:

| Dataset     |  Contain relu in last layer | Test | Validation |
| ----------- | ----------- |
| Ogbg-molhiv      | No       | 0.7964,       |0.7904|
|  Ogbg-molhiv | Yes       | 0.7922   | 0.8092 

| Ogbg-molpcba      | Yes       | 0.2288       |0.2308|
|  Ogbg-molpcba | No       | 0.2239   | 0.2283|